common/network-common/src/main/java/org/apache/spark/network/client/TransportResponseHandler.java

```diff
@@ -165,10 +175,37 @@ public void handle(ResponseMessage message) throws Exception {
       if (listener == null) {
         logger.warn("Ignoring response for block {} from {} since it is not outstanding",
           resp.streamChunkId, getRemoteAddress(channel));
-        resp.body().release();
       } else {
-        outstandingFetches.remove(resp.streamChunkId);
-        listener.onSuccess(resp.streamChunkId.chunkIndex, resp.body());
+        if (resp.isBodyInFrame()) {
+          outstandingFetches.remove(resp.streamChunkId);
+          listener.onSuccess(resp.streamChunkId.chunkIndex, resp.body());
+        } else {
+          StreamCallback<StreamChunkId> streamCallback =
+                  outstandingStreamFetches.get(resp.streamChunkId).get();
```

Comment:
the whole point of the supplier is to delay this call to `get()`, right?  you want to avoid creating the output file until here?

Reply:
Yes and to avoid creating unnecessary temp files for the small chunks  (under maxRemoteBlockSizeFetchToMem) which will be loaded into the memory.